<a href="https://colab.research.google.com/github/Pai-Ya-Ting/Text-mining/blob/main/Word%20Tokenization.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Assignment 1. Word Tokenization using Python

(1) Tokenization & Remove stopword without Stemming/Lemmatization.

(a) Read text from file carroll-alice.txt” and print it

In [ ]:
import urllib.request
url = 'https://raw.githubusercontent.com/rutum/tf-idf/master/gutenberg/carroll-alice.txt'
with urllib.request.urlopen(url) as f:
  text = f.read()
  print(text)

b'[Alice\'s Adventures in Wonderland by Lewis Carroll 1865]\n\nCHAPTER I. Down the Rabbit-Hole\n\nAlice was beginning to get very tired of sitting by her sister on the\nbank, and of having nothing to do: once or twice she had peeped into the\nbook her sister was reading, but it had no pictures or conversations in\nit, \'and what is the use of a book,\' thought Alice \'without pictures or\nconversation?\'\n\nSo she was considering in her own mind (as well as she could, for the\nhot day made her feel very sleepy and stupid), whether the pleasure\nof making a daisy-chain would be worth the trouble of getting up and\npicking the daisies, when suddenly a White Rabbit with pink eyes ran\nclose by her.\n\nThere was nothing so VERY remarkable in that; nor did Alice think it so\nVERY much out of the way to hear the Rabbit say to itself, \'Oh dear!\nOh dear! I shall be late!\' (when she thought it over afterwards, it\noccurred to her that she ought to have wondered at this, but at the time\nit a

(b) Normalize words: convert all upper-case letters to lower case (for example, 'Word' and 'word' are considered as the same word).

In [ ]:
lowtext = text.lower()
print(lowtext)

b'[alice\'s adventures in wonderland by lewis carroll 1865]\n\nchapter i. down the rabbit-hole\n\nalice was beginning to get very tired of sitting by her sister on the\nbank, and of having nothing to do: once or twice she had peeped into the\nbook her sister was reading, but it had no pictures or conversations in\nit, \'and what is the use of a book,\' thought alice \'without pictures or\nconversation?\'\n\nso she was considering in her own mind (as well as she could, for the\nhot day made her feel very sleepy and stupid), whether the pleasure\nof making a daisy-chain would be worth the trouble of getting up and\npicking the daisies, when suddenly a white rabbit with pink eyes ran\nclose by her.\n\nthere was nothing so very remarkable in that; nor did alice think it so\nvery much out of the way to hear the rabbit say to itself, \'oh dear!\noh dear! i shall be late!\' (when she thought it over afterwards, it\noccurred to her that she ought to have wondered at this, but at the time\nit a

(c) Tokenization: using nltk word_tokenize.

we converted the html above (a byte-like object) into a string using .decode and tokenized the content.

In [ ]:
import nltk  #nltk stand for natural language toolkit
nltk.download('punkt')
tokens = nltk.word_tokenize(lowtext.decode('utf-8'))
print(tokens)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
['[', 'alice', "'s", 'adventures', 'in', 'wonderland', 'by', 'lewis', 'carroll', '1865', ']', 'chapter', 'i.', 'down', 'the', 'rabbit-hole', 'alice', 'was', 'beginning', 'to', 'get', 'very', 'tired', 'of', 'sitting', 'by', 'her', 'sister', 'on', 'the', 'bank', ',', 'and', 'of', 'having', 'nothing', 'to', 'do', ':', 'once', 'or', 'twice', 'she', 'had', 'peeped', 'into', 'the', 'book', 'her', 'sister', 'was', 'reading', ',', 'but', 'it', 'had', 'no', 'pictures', 'or', 'conversations', 'in', 'it', ',', "'and", 'what', 'is', 'the', 'use', 'of', 'a', 'book', ',', "'", 'thought', 'alice', "'without", 'pictures', 'or', 'conversation', '?', "'", 'so', 'she', 'was', 'considering', 'in', 'her', 'own', 'mind', '(', 'as', 'well', 'as', 'she', 'could', ',', 'for', 'the', 'hot', 'day', 'made', 'her', 'feel', 'very', 'sleepy', 'and', 'stupid', ')', ',', 'whether', 'the', 'pleasure', 'of', 'm

(d) Remove stopwords using stopwords from nltk. You can add more stopword if needed.


Stop Words are words which do not contain important significance to be used in Search Queries. Usually, these words are filtered out from search queries because they return a vast amount of unnecessary information. 

In [ ]:
nltk.download('stopwords')
from nltk.corpus import stopwords
stopwords = set(stopwords.words('english'))
filtered_token = [word for word in tokens if not word in stopwords and word.isalnum()]
print(filtered_token)

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
['alice', 'adventures', 'wonderland', 'lewis', 'carroll', '1865', 'chapter', 'alice', 'beginning', 'get', 'tired', 'sitting', 'sister', 'bank', 'nothing', 'twice', 'peeped', 'book', 'sister', 'reading', 'pictures', 'conversations', 'use', 'book', 'thought', 'alice', 'pictures', 'conversation', 'considering', 'mind', 'well', 'could', 'hot', 'day', 'made', 'feel', 'sleepy', 'stupid', 'whether', 'pleasure', 'making', 'would', 'worth', 'trouble', 'getting', 'picking', 'daisies', 'suddenly', 'white', 'rabbit', 'pink', 'eyes', 'ran', 'close', 'nothing', 'remarkable', 'alice', 'think', 'much', 'way', 'hear', 'rabbit', 'say', 'dear', 'oh', 'dear', 'shall', 'late', 'thought', 'afterwards', 'occurred', 'ought', 'wondered', 'time', 'seemed', 'quite', 'natural', 'rabbit', 'actually', 'took', 'watch', 'looked', 'hurried', 'alice', 'started', 'feet', 'flashed', 'across', 'mind', 'ne

The function would print the occurrence of words and show the top 10 words.

In [ ]:
def term_count(token_doc):
  text_tf = dict((term, token_doc.count(term)) for term in token_doc)
  print(text_tf)
  freq = nltk.FreqDist(text_tf)
  print("Top 10 freqency word", freq.most_common(10))
  return freq.most_common(10)

(e) Count the occurrence of words and display the top 10 common words.

In [ ]:
initial = term_count(filtered_token)

{'alice': 396, 'adventures': 7, 'wonderland': 3, 'lewis': 1, 'carroll': 1, '1865': 1, 'chapter': 12, 'beginning': 14, 'get': 44, 'tired': 7, 'sitting': 10, 'sister': 9, 'bank': 3, 'nothing': 30, 'twice': 5, 'peeped': 3, 'book': 7, 'reading': 3, 'pictures': 4, 'conversations': 1, 'use': 18, 'thought': 74, 'conversation': 10, 'considering': 3, 'mind': 11, 'well': 37, 'could': 86, 'hot': 5, 'day': 25, 'made': 30, 'feel': 8, 'sleepy': 5, 'stupid': 3, 'whether': 11, 'pleasure': 2, 'making': 8, 'would': 90, 'worth': 4, 'trouble': 6, 'getting': 22, 'picking': 2, 'daisies': 1, 'suddenly': 13, 'white': 30, 'rabbit': 47, 'pink': 1, 'eyes': 28, 'ran': 16, 'close': 13, 'remarkable': 2, 'think': 53, 'much': 51, 'way': 53, 'hear': 14, 'say': 51, 'dear': 28, 'oh': 22, 'shall': 23, 'late': 6, 'afterwards': 2, 'occurred': 2, 'ought': 14, 'wondered': 1, 'time': 68, 'seemed': 27, 'quite': 55, 'natural': 4, 'actually': 1, 'took': 24, 'watch': 8, 'looked': 45, 'hurried': 11, 'started': 2, 'feet': 19, 'flas


2. Tokenization & Remove stopwords with Stemming/Lemmatization.

PorsterStemmer

Stemming is the process of reducing inflection in words to their root forms such as mapping a group of words to the same stem even if the stem itself is not a valid word in the Language.

In [ ]:
from nltk import PorterStemmer
ps = PorterStemmer()
token_psstem = [ps.stem(word) for word in filtered_token]

stem = term_count(token_psstem)

{'alic': 396, 'adventur': 7, 'wonderland': 3, 'lewi': 1, 'carrol': 1, '1865': 1, 'chapter': 12, 'begin': 30, 'get': 66, 'tire': 7, 'sit': 17, 'sister': 11, 'bank': 4, 'noth': 30, 'twice': 5, 'peep': 4, 'book': 8, 'read': 13, 'pictur': 6, 'convers': 11, 'use': 35, 'thought': 80, 'consid': 7, 'mind': 13, 'well': 37, 'could': 86, 'hot': 5, 'day': 29, 'made': 30, 'feel': 17, 'sleepi': 5, 'stupid': 3, 'whether': 11, 'pleasur': 2, 'make': 46, 'would': 90, 'worth': 4, 'troubl': 6, 'pick': 5, 'daisi': 1, 'suddenli': 13, 'white': 38, 'rabbit': 48, 'pink': 1, 'eye': 35, 'ran': 16, 'close': 16, 'remark': 28, 'think': 63, 'much': 54, 'way': 54, 'hear': 18, 'say': 70, 'dear': 31, 'oh': 22, 'shall': 23, 'late': 7, 'afterward': 2, 'occur': 2, 'ought': 14, 'wonder': 28, 'time': 74, 'seem': 40, 'quit': 55, 'natur': 4, 'actual': 1, 'took': 24, 'watch': 13, 'look': 102, 'hurri': 23, 'start': 2, 'feet': 19, 'flash': 1, 'across': 5, 'never': 45, 'seen': 15, 'either': 9, 'take': 27, 'burn': 3, 'curios': 5, 

Lemmatization

Lemmatization reduces the inflected words properly ensuring that the root word belongs to the language.

In [ ]:
nltk.download('wordnet')
from nltk.stem import WordNetLemmatizer

wnl = WordNetLemmatizer()
token_lem = [wnl.lemmatize(word) for word in filtered_token]

lemma = term_count(token_lem)

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
{'alice': 396, 'adventure': 7, 'wonderland': 3, 'lewis': 1, 'carroll': 1, '1865': 1, 'chapter': 12, 'beginning': 14, 'get': 44, 'tired': 7, 'sitting': 10, 'sister': 11, 'bank': 4, 'nothing': 30, 'twice': 5, 'peeped': 3, 'book': 8, 'reading': 3, 'picture': 5, 'conversation': 11, 'use': 18, 'thought': 76, 'considering': 3, 'mind': 11, 'well': 37, 'could': 86, 'hot': 5, 'day': 29, 'made': 30, 'feel': 8, 'sleepy': 5, 'stupid': 3, 'whether': 11, 'pleasure': 2, 'making': 8, 'would': 90, 'worth': 4, 'trouble': 6, 'getting': 22, 'picking': 2, 'daisy': 1, 'suddenly': 13, 'white': 30, 'rabbit': 48, 'pink': 1, 'eye': 35, 'ran': 16, 'close': 13, 'remarkable': 2, 'think': 53, 'much': 51, 'way': 54, 'hear': 14, 'say': 55, 'dear': 31, 'oh': 22, 'shall': 23, 'late': 6, 'afterwards': 2, 'occurred': 2, 'ought': 14, 'wondered': 1, 'time': 74, 'seemed': 27, 'quite': 55, 'natural': 4, 'actuall

3. Compare the result (top 10 common words) of question 1 (without stemming/ Lemmatization) 
and question 2 (with stemming and with Lemmatization).
Are the results different? 
Try to explain why they’re different, or give your opinion about the result.
In your opinion, which way is the best?

Stemming a word or sentence may result in words that are not actual words. Stems are created by removing the suffixes or prefixes used with a word without knowledge of the context and therefore can't discriminate between words which have different meanings depending on part of speech.

Lemmatization returns an actual word of the language, it is used where it is necessary to get valid words.
In lemmatisation, the part of speech of a word should be first determined and the normalisation rules will be different for different part of speech.

The difference between them explains why 'look' appears in the result of stemming but not in the result of lemmatization.

For example, after stemming, the term 'queer-looking' would be counted as 'look'. However, if we do lemmatization, it won't be counted. Therefore, 'look' appears 102 times in the result with stemming  while 'looked' only appears 45 times in the one with lemmatization.

The difference before/after lemmatization is that it identify a canonical representative for a set of related word forms.

Take 'take' for example, the result without lemmatization would see 'takes' and 'taken' as different words.

In conclusion, lemmatization offers better precision than stemming. However, if the tenses do matter, both of them shouldn't be used.

In [ ]:
print(initial)
print(stem)
print(lemma)

[('said', 462), ('alice', 396), ('little', 128), ('one', 99), ('would', 90), ('know', 88), ('could', 86), ('like', 85), ('went', 83), ('queen', 75)]
[('said', 462), ('alic', 396), ('littl', 128), ('look', 102), ('one', 100), ('like', 97), ('know', 92), ('would', 90), ('could', 86), ('went', 83)]
[('said', 462), ('alice', 396), ('little', 128), ('one', 100), ('would', 90), ('know', 90), ('could', 86), ('like', 86), ('went', 83), ('thing', 79)]
